shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -230,35 +231,11 @@ public boolean sendKeyEvent(KeyEvent event) {
           updateEditingState();
           return true;
         } else if (selStart > 0) {
-          // Delete to the left/right of the cursor depending on direction of text.
-          // TODO(garyq): Explore how to obtain per-character direction. The
-          // isRTLCharAt() call below is returning blanket direction assumption
-          // based on the first character in the line.
-          boolean isRtl = mLayout.isRtlCharAt(mLayout.getLineForOffset(selStart));
-          try {
-            if (isRtl) {
-              Selection.extendRight(mEditable, mLayout);
-            } else {
-              Selection.extendLeft(mEditable, mLayout);
-            }
-          } catch (IndexOutOfBoundsException e) {
-            // On some Chinese devices (primarily Huawei, some Xiaomi),
-            // on initial app startup before focus is lost, the
-            // Selection.extendLeft and extendRight calls always extend
-            // from the index of the initial contents of mEditable. This
-            // try-catch will prevent crashing on Huawei devices by falling
-            // back to a simple way of deletion, although this a hack and
-            // will not handle emojis.
-            Selection.setSelection(mEditable, selStart, selStart - 1);
+          if (TextKeyListener.getInstance().onKeyDown(null, mEditable, event.getKeyCode(), event)) {
```

Comment:
I tested with a few keyboard, and it looks like this isn't quite working with extended grapheme cluster emojis, such as flag emojis: Type a few flag emojis, and try to delete them from the middle (move caret to middle, then delete) On a few of my devices, I am seeing the flag emoji only partially delete and turn into its components.

Reply:
It looks like the same happens with the Qwerty version of this.

Reply:
Does it still happen without this PR?  On master I've noticed I can turn a grapheme cluster emoji into its components using the arrow keys to put the cursor in the middle.

Reply:
I think the problem is another bug in moving cursor around. When touching TextField on a real device I got no bugs. But when dragging it (sometimes) or moving with keyboard arrows, cursor is not placed correctly.
When the cursor place was correct, my code worked.
Did you experience the problem on real devices and touch interfaces?
If yes, give me more information, please

Reply:
Yes, this was all on real devices with touch.

Reproed on:
SM-T380 (Galaxy Tab)
Pixel 3
Samsung Galaxy Note 9
Huawei Nova 4

Here is a screenshot:
![flutter_01](https://user-images.githubusercontent.com/1887398/78728191-8b6c4e00-78eb-11ea-803e-3a34ff14c3fb.png)

Directions: Create app with basic TextField widget
Type any flag emojis
Tap in the middle, caret will appear correctly between flags.
Press backspace. The flag emoji to be deleted will break apart, with the letter emoji and a question mark representing the zero width joiner.

The reason is that the key listener is deleting by codepoint. This is fine for simple emojis, but will break more complex emojis such as family emojis which depend on many component emojis linked with "zero width joiner" characters in between. The goal of the expandLeft and expandRight is that those android methods would expand by glyph, anutomatically handling any codepoints that would be encountered. The bug is caused by RTL scenarios mixing with LTR characters, which may appear on the visually wrong side to be encountered by the expand, which expands visually.

Reply:
I think caret place is visually correct. I'm currently working on a better solution or at least a hotfix for RTL scenarios

Reply:
Are you sure this bug doesn't happen without this PR?

Reply:
The reason I say the problem might be with caret place is that when the caret was in the end or when I placed it correctly with arrow keys, the code could delete a complete glyph.